### PR TITLE
Tap element2

### PR DESCRIPTION
**Pull request by BryceStevenWilley**
Found a decent solution to tapping tabbed templates, and for tapping arbitrary elements. Not urgent, so leaving this in drafts until I can iterate on it for a bit longer.

### CODE REVIEW

**Comment by BryceStevenWilley** [comment]
I moved over some fixes from the accessibility branch that I had working on recently, but this should be ready for review. I'll make sure the tests are passing before pinging anyone there.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js:253]: I think you're right.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js:247]: I've been making the second argument an object. That does complicate factors in some way, but it may be worth talking about at some point. Could you add a description of what the expected value is?

Comment (plocket) [reply lib/scope.js:247]: Or maybe it just could have a different name, like `tab_id`.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js:277]: This [already exists in v7 of cucumber](https://github.com/cucumber/cucumber-js/blob/7.x/docs/support_files/api_reference.md), a version which we use, we just haven't made the conversion yet. That said, if I recall correctly some of this doesn't actually belong after every step (hence the various arguments) and we have to think more carefully about how to handle this functionality.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js]: Just noting that here too `waitForTab` might be `tab_id` or `tab_id_to_wait_for` or `tab_to_wait_for`. `waitForTab` currently sounds like a boolean to me. That said, this isn't a blocking change.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js]: Is this a good place for these comments? I'm not sure the previous place was the best place for all of these, but I think it may have been slightly better.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js]: Just personal confusion experience - what is the name `inner` meant to describe?

File: lib/scope.js
Comment (plocket) [inline lib/scope.js]: Might change this name to `tabElementBySelector` since it really can be any selector, right? It would also let people know that they have to include `#` before the selector if they want to us an id - this isn't going to add it automatically.

File: lib/steps.js
Comment (plocket) [inline lib/steps.js:393]: I like it!

File: lib/steps.js
Comment (plocket) [inline lib/steps.js]: Why have a hard-coded timeout? Does it make sense to use the timeout the user can set themselves? That is, `scope.timeout`?

File: lib/scope.js
Comment (plocket) [inline lib/scope.js]: Just a style note - `.evaluate()` returns a promise, so you shouldn't need to use `Promise.all()` here.

File: lib/scope.js
Comment (plocket) [reply lib/scope.js:277]: Also, some of this should be done while the story table runs between pages, etc, and a story table is only one Step, so we'll have to manage that manually anyway.

Not easy stuff to know about this system. I'm not yet sure how to clarify it, but a name change would probably help.

**Review by plocket — APPROVED** [review]
Great addition! None of these comments are blocking changes, so I'll approve, but maybe we can make some into issues. I think the two main concerns that I noted are:

First, the hard-coded timeout in the `tap "..." element`. Even though there's a way to have a custom timeout, I still don't see why we'd have a hidden hard-coded timeout.

Second, the name of `tapElementById`, which I think is a misleading name.

Thanks for all the work!

**Comment by plocket** [comment]
Actually, is there a way to test any of these at the moment? [We can make that into an issue for the future, but we do want to at least note that.]

Edit: Also, let me know if you have bandwidth to make an [issue in the documentation repo](https://github.com/suffolkLITLab/docassemble-assemblyline-documentation/issues) for the various new steps.

File: lib/steps.js
Comment (BryceStevenWilley) [reply lib/steps.js]: That's a separately defined step (that's just a different interface to this same code) that lets someone specify a timeout. IMO I didn't want to specify the timeout for every single step that I used this tap feature in, and unfortunately we do still need the time out. It's for browser events, so I can't imagine it being too flaky, since it's not based on server events. Ideally, the author / running shouldn't have to know that this timeout exists, so I don't want to make it super configurable elsewhere.

I looked into possibly using the [transition to end](https://stackoverflow.com/questions/51547994/wait-for-transition-to-end-in-puppeteer) value, but I couldn't actually get it to work for the one example I had (tapping on tabs in the tabbed interface), and I think what the actual transition that we should wait for will vary and I can't actually find?

**Comment by BryceStevenWilley** [comment]
> Actually, is there a way to test any of these at the moment?

You're right, there's nothing stopping that from being added to the test repo (other than the friction of doing so). I'll do that and ping back to this PR when it's ready.

Can also add things to the documentation repo too, just wanted a confirmation that this is a wanted feature for the main repo.

File: lib/steps.js
Comment (plocket) [reply lib/steps.js]: I'm willing to go with that choice.

Regarding catching transition events, you can go down the rabbit hole if you want, but it depends on the implementation being used and, yes, it would be specific to the situation. Also, tapping some things may not cause transitions, etc. It's pretty complicated even before you hit the difficulties in the code itself.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js:253]: This is straight copied from the above function, I believe you wrote this, haha.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js:247]: So it is the tab id, but not being used to get the tab, it's the labelledby text on the tab panel (where the contents are actually shown, not the item you click to show it, which is the tab). I'll add this description, don't think there's a better name that's not extremely long and unwieldy though.

I'm going to leave the arguments as is too, I don't think it needs that many options right now.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js]: It was a lazy name: it was the inner promise of the promise race below. I've changed it to be a bit more clear.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js]: It's in almost the same place? It was inside the promise.all below, but I had to duplicate that call, so it can't (and shouldn't) stay in the exact same place because it applies both to lines 321 and 326.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js]: This function name isn't something that authors will see, but I like the new name anyway. Changed.

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js:277]: A lot of that is out of scope for this PR: I just corrected a typo when I saw it.

File: lib/steps.js
Comment (BryceStevenWilley) [reply lib/steps.js]: I ran into some problems when testing this to make sure it worked, and ended up trying the transition events and it seems to work pretty well? I want to see if it'll work for the accessibility testing before merging (I remembered that the main reason I needed the wait was because I would tap the button to switch the tabs and then check for accessibility, but the page wouldn't have fully transitioned yet, so the accessibility check would check the colors with the in-progress fade-in. Those colors didn't have enough contrast, but also weren't the true colors of the page and so would be a flaky failure), but it's pretty smooth right now.

**Comment by BryceStevenWilley** [comment]
Test interview file in https://github.com/plocket/docassemble-ALAutomatedTestingTests/pull/187. Tested locally and passing 100%.

File: lib/steps.js
Comment (BryceStevenWilley) [reply lib/steps.js]: Accessibility seems to work fine with the transition end stuff as well. I think it should be good.

File: lib/scope.js
Comment (plocket) [reply lib/scope.js]: It's a tad complex, so we can probably leave it as is for now. Here are thoughts to maybe make into an issue for future clean up.

Line 311: Think it goes under Promise.race, since it explains (I think?) why we're not racing to detect an invalid input error.

Line 312: Probably goes with the Promise that clicks and navigates.

Lines 313-319: not sure where they should go, so we can leave them here, but they should go somewhere else someday if we can keep an eye out for that. Maybe in our own repo's docs.

File: lib/scope.js
Comment (plocket) [reply lib/scope.js]: I think maybe you didn't push that change.

File: lib/steps.js
Comment (plocket) [reply lib/steps.js]: I'll have to see the code to see if it's general enough. To use it here, this would have to be able to detect the transition of tapping any element, not just a tab. If you want to just detect tab tapping transitions, that should probably be a separate Step just for tabs.

Comment (plocket) [reply lib/steps.js]: Somehow my page wasn't updating with new code. I'll check it out now and see how it all flows.

File: lib/scope.js
Comment (plocket) [reply lib/scope.js]: Page hadn't updated, looks good to me!

File: lib/scope.js
Comment (BryceStevenWilley) [reply lib/scope.js]: I moved those first two lines to be in the right places. I think the other lines are fine where they are

File: tests/features/interactive_steps.feature
Comment (plocket) [inline tests/features/interactive_steps.feature]: Future issue: Would waiting 1 second be enough? I don't see a need for 5 and our tests are going to continue growing. I'd rather keep the time we add manually to a minimum.

Doesn't need to block this PR.

File: lib/scope.js
Comment (plocket) [inline lib/scope.js:488]: Nice find!

Comment (BryceStevenWilley) [reply lib/scope.js:488]: Thanks! I don't think it's necessary as a comment, but I found it from trying to debug some of the puppeteer `evaluate` functions: https://stackoverflow.com/a/49887254/11416267.

File: tests/features/interactive_steps.feature
Comment (BryceStevenWilley) [reply tests/features/interactive_steps.feature]: I thought I had a reason for 5, but 1 second seems to work fine locally at least. I changed it to 1 (and added another fix to the step regex to allow for grammatically correct steps), hopefully it's not too flaky.

**Comment by plocket** [comment]
Future: I think one remaining improvement we could make in the future is to better handle when the tab doesn't exist. For example, for when the user has a typo.

File: lib/scope.js
Comment (plocket) [reply lib/scope.js:488]: Just under that I see a link to https://github.com/puppeteer/puppeteer#debugging-tips with more tips.

**Comment by BryceStevenWilley** [comment]
> Future: I think one remaining improvement we could make in the future is to better handle when the tab doesn't exist. For example, for when the user has a typo.

How you you suggest handling it? `waitForSelector` just waits for `scope.timeout` seconds and if it's not there returns null, which will fail later. Would it just be an additional check for null afterwards and a better log message? If that's all, I'll do it now.

https://puppeteer.github.io/puppeteer/docs/puppeteer.page.waitforselector/

**Comment by plocket** [comment]
> How you you suggest handling it?

I think I'd try to detect the tab that has the actual ID. If it exists, can we assume its content exists too?

**Comment by plocket** [comment]
I suggest not using `waitForSelector` as that will slow everything down. Unless there's a reason you can think that the element would take a while to appear on the page. [Use `.$()` instead.]

[That is, don't check afterwards, check beforehand more quickly.]

**Comment by BryceStevenWilley** [comment]
Everything should be good (addressed the last comment, and added info to the changelog). The documentation is ready in https://github.com/SuffolkLITLab/docassemble-AssemblyLine-documentation/pull/244 and will be merged soon. I'll squash and merge now.

**Pull request merged**
